pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -55,268 +131,477 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+
+	return reconcile.Result{}, nil
 }
 
-var Handler ReconcileHandler
+func (r *ReconcileNodeMaintenance) processMaintModeOff(statusAnnoData *ReconcileNodeMaintenanceStatusInfo, node *corev1.Node) (reconcile.Result, error) {
 
-// Reconcile reads that state of the cluster for a NodeMaintenance object and makes changes based on the state read
-// and what is in the NodeMaintenance.Spec
-// Note:
-// The Controller will requeue the Request to be processed again if the returned error is non-nil or
-// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
-func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconcile.Result, error) {
-	reqLogger := log.WithFields(log.Fields{"Request.Namespace": request.Namespace, "Request.Name": request.Name})
-	reqLogger.Info("Reconciling NodeMaintenance")
+	nodeName := node.ObjectMeta.Name
 
-	// Fetch the NodeMaintenance instance
-	instance := &kubevirtv1alpha1.NodeMaintenance{}
-	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	drainer, _, err := r.initDrainer()
 	if err != nil {
-		if errors.IsNotFound(err) {
-			// Request object not found, could have been deleted after reconcile request.
-			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
-			// Return and don't requeue
-			reqLogger.Infof("NodeMaintenance Object: %s Deleted ", request.NamespacedName)
-			return reconcile.Result{}, nil
-		}
-		// Error reading the object - requeue the request.
-		reqLogger.Info("Error reading the request object, requeuing.")
-		return reconcile.Result{}, err
+		return reconcile.Result{Requeue: true, RequeueAfter: RequeuDrainingWaitTime * time.Second}, nil
 	}
 
-	// Add finalizer when object is created
-	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
-		if !ContainsString(instance.ObjectMeta.Finalizers, kubevirtv1alpha1.NodeMaintenanceFinalizer) {
-			instance.ObjectMeta.Finalizers = append(instance.ObjectMeta.Finalizers, kubevirtv1alpha1.NodeMaintenanceFinalizer)
-			if err := r.client.Update(context.TODO(), instance); err != nil {
-				return r.reconcileAndError(instance, err)
-			}
-		}
-	} else {
-		// The object is being deleted
-		if ContainsString(instance.ObjectMeta.Finalizers, kubevirtv1alpha1.NodeMaintenanceFinalizer) {
-			// Stop node maintenance - uncordon and remove live migration taint from the node.
-			if err := r.stopNodeMaintenance(instance.Spec.NodeName); err != nil {
-				if errors.IsNotFound(err) == false {
-					return r.reconcileAndError(instance, err)
-				}
-			}
-			// Remove our finalizer from the list and update it.
-			instance.ObjectMeta.Finalizers = RemoveString(instance.ObjectMeta.Finalizers, kubevirtv1alpha1.NodeMaintenanceFinalizer)
-			if err := r.client.Update(context.Background(), instance); err != nil {
-				return r.reconcileAndError(instance, err)
-			}
+	if err := r.cancelEviction(drainer, node, statusAnnoData); err != nil {
```

Comment:
What if another controller starts evicting Pods after the maintenance window ends?

Reply:
should i report an error  if an cordoning is already set when the user requests us to enter maintenance mode?

Reply:
What other piece of information could you use to decide whether to call ``cancelEviction()``?